opal/mca/btl/tcp/btl_tcp_component.c

```diff
@@ -1299,45 +1338,156 @@ static void mca_btl_tcp_component_recv_handler(int sd, short flags, void* user)
     struct sockaddr_storage addr;
     opal_socklen_t addr_len = sizeof(addr);
     mca_btl_tcp_proc_t* btl_proc;
-    int retval;
+    bool sockopt = true;
+    size_t retval, len = strlen(mca_btl_tcp_magic_id_string);
+    mca_btl_tcp_endpoint_hs_msg_t hs_msg;
+    struct timeval save, tv;
+    socklen_t rcvtimeo_save_len = sizeof(save);
+    char str[128];
+
+    /* Setting SO_RCVTIMEO to say 2 seconds here to avoid chance of spin
+     * forever if it tries to connect to old version
+     * as older version will send just process id which won't be long enough
+     * to cross sizeof(str) length + process id struct*/
```

Comment:
Just to hit those of us who aren't that smart over the head a bit more, I'd add to this comment that you leave the socket in blocking mode until the initial receive is complete, so that RCVTIMEO does something sane.  Possibly explain the usage of RCvTIMEO (to handle the case where the remote side isn't OMPI and isn't going to send us any data).  The why is important :).